### PR TITLE
Feature/855 update settings dashboard

### LINKED ISSUE #855: Create settings page which allows admins to change the hackathon settings
We have a settings schema on API-side, but we don't use it on the dashboard. Let's add a page that allows admins to update this.

### PR DESCRIPTION
**Pull request by pierreTklein**
### Tickets:

fixes #855

### List of changes:

- Create settings page which allows admins to change the open, close, and confirm times for hackers
- Add functions into util file

### Type of change:

Please delete options that aren't relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

### How to test:

- Open dashboard, navigate to settings page
- Change the date / time
- Confirm on reload that the new date / time is returned by the api 

### Questions:

- I think this will fail if there does not exist a settings schema in the DB. I will need to check this, but if you want, please confirm.

### PR Checklist:

- [ ] Merged `develop` branch (before testing)
- [x] Linted my code locally
- [ ] Listed change(s) in the Changelog
- [x] Tested all links in project relevant browsers
- [x] Tested all links on different screen sizes
- [x] Referenced all useful info (issues, tasks, etc)

### Screenshots:

Pretty bare-bones. Not looking to make it nice atm, just getting it functional.
image

Confirmation modal:
[Image from Gyazo](https://gyazo.com/dd847c8058308f51b3f72b6f6416e9dc)

### CODE REVIEW

**Comment by krubenok** [comment]
Ok it's not that I don't trust the admins, but I also don't trust the admins lol. Everyone screws up and having this in UX exposes more people to it in a more innocent mistake kinda way. 

I think there needs to be some UX here to be like "Hey are you double triple super certain you wanna do this?"

File: src/features/Settings/SettingsContainer.tsx [hunk 1]
```diff
@@ -0,0 +1,64 @@
+import { Flex } from '@rebass/grid';
```
Comment (logan-r) [inline src/features/Settings/SettingsContainer.tsx:1]: This whole file should be moved into `/pages/Admin/Settings.tsx`

File: src/util/util.ts [hunk 2]
```diff
@@ -45,6 +45,38 @@ function date2input(date: string) {
   return `${month}${day}${year}`;
 }
 
+function datetime2input(date: string) {
```
Comment (logan-r) [inline src/util/util.ts:48]: Could you add a comment here detailing format of input and output?

Same for `input2datetime`

**Comment by logan-r** [comment]
I agree with @krubenok that we need to have some sort of "confirm you are absolutely sure you want to change this" check
As for implementation a `window.confirm` popup might be enough or maybe something more directly integrated into the form, like forcing them to type dates twice à la password update?

File: src/features/Settings/SettingsContainer.tsx (hunk 1 above)
Comment (pierreTklein) [reply src/features/Settings/SettingsContainer.tsx]: Done

File: src/pages/Admin/Settings.tsx [hunk 3]
```diff
@@ -1,7 +1,82 @@
...
+import { Settings } from '../../api/settings';
+import { HACKATHON_NAME, ISetting } from '../../config';
+import SettingsForm from '../../features/Settings/SettingsForm';
+import { ConfirmModal, H1, MaxWidthBox } from '../../shared/Elements';
+import ValidationErrorGenerator from '../../shared/Form/validationErrorGenerator';
+import WithToaster from '../../shared/HOC/withToaster';
+import theme from '../../shared/Styles/theme';
 
-import SettingsContainer from '../../features/Settings/SettingsContainer';
+const SettingsContainer: React.FC = () => {
```
Comment (logan-r) [inline src/pages/Admin/Settings.tsx:13]: nit:
should be `SettingsPage`

File: src/features/Settings/SettingsContainer.tsx (hunk 1 above)
Comment (logan-r) [reply src/features/Settings/SettingsContainer.tsx]: Cheers 👍

File: .vscode/settings.json [hunk 4]
```diff
@@ -2,5 +2,5 @@
   "editor.formatOnSave": true,
   "files.insertFinalNewline": true,
   "files.trimFinalNewlines": true,
-  "editor.tabSize": 4
+  "editor.tabSize": 2
```
Comment (loreina) [inline .vscode/settings.json:5]: we'll have to run `format` on the whole repo for this to work in full effect I think?

File: src/util/util.ts (hunk 2 above)
Comment (pierreTklein) [reply src/util/util.ts]: Done

File: src/config/constants.ts [hunk 5]
```diff
@@ -135,3 +135,8 @@ export const TEAM_MAX_SIZE = 'Max 4 Hackers';
 export const SPONSOR_COMPANY_LABEL = 'Company name';
 export const SPONSOR_CONTRACT_URL_LABEL = 'Contract URL';
 export const SPONSOR_NOMINEE_LABEL = 'Nominees';
+
+// Settings management page
+export const SETTINGS_OPEN_TIME_LABEL = 'Applications Open At';
+export const SETTINGS_CLOSE_TIME_LABEL = 'Applications Close At';
+export const SETTINGS_CONFIRM_TIME_LABEL = 'Hacker Confirmations Close At';
```
Comment (loreina) [inline src/config/constants.ts:142]: nit: sentence case if it's more than just a label of the field (i.e. First Name vs What's your first name? need better docs on ux copy tho)

```suggestion
export const SETTINGS_OPEN_TIME_LABEL = 'Applications open at:';
export const SETTINGS_CLOSE_TIME_LABEL = 'Applications close at:';
export const SETTINGS_CONFIRM_TIME_LABEL = 'Hacker confirmations close at:';
```

File: src/pages/Admin/Settings.tsx (hunk 3 above)
Comment (pierreTklein) [reply src/pages/Admin/Settings.tsx:13]: Done

File: .vscode/settings.json (hunk 4 above)
Comment (pierreTklein) [reply .vscode/settings.json:5]: Yeah. This is already in .prettierrc. Maybe we will want to do the full change in a later PR?

Comment (loreina) [reply .vscode/settings.json:5]: yeah wanna remove for now? might be prettier that's inconsistent b/c hackerAPI config is `tabWidth: 4` for vscode and prettier

Comment (pierreTklein) [reply .vscode/settings.json:5]: Do you mean to say change prettierrc to be 4? Or leave the inconsistency?

Comment (pierreTklein) [reply .vscode/settings.json]: Done

**Pull request merged**